### PR TITLE
Add various loads and stores to JitIL

### PR DESCRIPTION
**Pull request by MoochMcGee**
I've added lhau, lhaux, lfsu, and lfdu. Also, because stfd already implements stfdu, I updated the table to reflect this.

### CODE REVIEW

**Comment by lioncash** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/Core/PowerPC/JitILCommon/JitILBase_LoadStoreFloating.cpp [hunk 1]
```diff
@@ -25,6 +25,21 @@ void JitILBase::lfs(UGeckoInstruction inst)
...
+	FALLBACK_IF(js.memcheck);
+
+	IREmitter::InstLoc addr = ibuild.EmitIntConst(inst.SIMM_16), val;
+
+	addr = ibuild.EmitAdd(addr, ibuild.EmitLoadGReg(inst.RA));
+
+	val = ibuild.EmitDupSingleToMReg(ibuild.EmitLoadSingle(addr));
+	ibuild.EmitStoreFReg(val, inst.FD);
+	ibuild.EmitStoreGReg(addr, inst.RA);
+	return;
```
Comment (lioncash) [inline Source/Core/Core/PowerPC/JitILCommon/JitILBase_LoadStoreFloating.cpp]: return isn't necessary here.

Comment (MoochMcGee) [inline Source/Core/Core/PowerPC/JitILCommon/JitILBase_LoadStoreFloating.cpp]: Ah, okay. It was in the original.

**Comment by lioncash** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/Core/PowerPC/JitILCommon/JitILBase_LoadStoreFloating.cpp [hunk 2]
```diff
@@ -21,10 +21,23 @@ void JitILBase::lfs(UGeckoInstruction inst)
...
+{
+	INSTRUCTION_START
+	JITDISABLE(bJITLoadStoreFloatingOff);
+	FALLBACK_IF(js.memcheck);
+
+	IREmitter::InstLoc addr = ibuild.EmitIntConst(inst.SIMM_16), val;
+
+	addr = ibuild.EmitAdd(addr, ibuild.EmitLoadGReg(inst.RA));
+
+	val = ibuild.EmitDupSingleToMReg(ibuild.EmitLoadSingle(addr));
```
Comment (lioncash) [inline Source/Core/Core/PowerPC/JitILCommon/JitILBase_LoadStoreFloating.cpp]: You can just declare this as assignment via: `IREmitter::InstLoc val = ibuild.EmitDupSingleToMReg(ibuild.EmitLoadSingle(addr));`

File: Source/Core/Core/PowerPC/JitILCommon/JitILBase_LoadStoreFloating.cpp [hunk 3]
```diff
@@ -42,6 +55,21 @@ void JitILBase::lfd(UGeckoInstruction inst)
 	ibuild.EmitStoreFReg(val, inst.RD);
 }
 
+void JitILBase::lfdu(UGeckoInstruction inst)
+{
+	INSTRUCTION_START
+	JITDISABLE(bJITLoadStoreFloatingOff);
+	FALLBACK_IF(js.memcheck);
+
+	IREmitter::InstLoc addr = ibuild.EmitIntConst(inst.SIMM_16), val;
```
Comment (lioncash) [inline Source/Core/Core/PowerPC/JitILCommon/JitILBase_LoadStoreFloating.cpp]: Might want to put `val` on the next line as a separate declaration.

**Comment by magumagu** [comment]
@dolphin-emu-bot rebuild

**Comment by magumagu** [comment]
It would be nice if the load instructions could be refactored a bit, but the implementations look fine.

Do we have hwtest coverage for these instructions?

**Comment by lioncash** [comment]
I believe we have partial tests in the hwtests branch.

**Comment by MoochMcGee** [comment]
Yes, @magumagu, they pass on this branch, I've checked.

**Comment by phire** [comment]
This PR causes F-Zero GX to generate a "Retrieving unknown spill slot?!" warning just before showing the intro sequence.

**Comment by MoochMcGee** [comment]
@phire Odd. Can you try disabling instructions in the JitIL to see which one's the culprit?

**Comment by phire** [comment]
lfsu appears to be the culprit.

**Comment by MoochMcGee** [comment]
Alright, I'll check it out in the morning.

**Comment by MaJoRoesch** [comment]
The [forum post reporting this error](https://forums.dolphin-emu.org/Thread-jitil-is-completly-broken-in-4-0-5962) also reported that Double Dash (the lakitu fishing rod) and Super Smash Bros Melee (any environment with the 3D camera (i.e. a fight)) have it too. When testing them with panic handlers on, they show the same error that F-Zero GX does, so it's probably the same instruction.

**Comment by MoochMcGee** [comment]
Ah, okay. I had noticed some problems with Melee.

**Comment by MaJoRoesch** [comment]
I went on ahead and made an issue report for this: [Issue 8437](https://code.google.com/p/dolphin-emu/issues/detail?id=8437).

**Comment by phire** [comment]
I tracked the problem deep into JitIL, where it mishandles the following IR:

1. imm32
2. reg
3. add 2 1
4. LoadSingle 3 (Load/Store double/single are effected)
5. use 3 (Any instruction which uses the result of the add in anyway)

The regBuildMemAddress function in LoadSingle preforms an optimisation if it's operand is and add, it ignores the add and builds a load instruction using the add's operands directly.

But the dead register tracking fails and the LoadSingle instruction kills both the register at 2 and the result from the add instruction at 3. When the instruction at 5 goes to use the result of the add instruction, it's nolonger in any register so it assumes it has been spilled and emits an error when it doesn't find a spilled register.

**Comment by phire** [comment]
The immediate issue should be fixed in #2272.

**Pull request merged**